hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/DirectInStoreCompactor.java

```diff
@@ -0,0 +1,90 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hbase.regionserver.compactions;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hbase.io.compress.Compression;
+import org.apache.hadoop.hbase.io.hfile.CacheConfig;
+import org.apache.hadoop.hbase.io.hfile.HFileContext;
+import org.apache.hadoop.hbase.regionserver.HStore;
+import org.apache.hadoop.hbase.regionserver.HStoreFile;
+import org.apache.hadoop.hbase.regionserver.StoreFileWriter;
+import org.apache.yetus.audience.InterfaceAudience;
+
+@InterfaceAudience.Private
+public class DirectInStoreCompactor extends DefaultCompactor {
+  public DirectInStoreCompactor(Configuration conf, HStore store) {
+    super(conf, store);
+  }
+
+  @Override
+  protected StoreFileWriter initWriter(FileDetails fd, boolean shouldDropBehind, boolean major)
+    throws IOException {
+    // When all MVCC readpoints are 0, don't write them.
+    // See HBASE-8166, HBASE-12600, and HBASE-13389.
+    return createWriterInFamilyDir(fd.maxKeyCount,
+      major ? majorCompactionCompression : minorCompactionCompression,
+      fd.maxMVCCReadpoint > 0, fd.maxTagsLength > 0, shouldDropBehind);
+  }
+
+  private StoreFileWriter createWriterInFamilyDir(long maxKeyCount,
+      Compression.Algorithm compression, boolean includeMVCCReadpoint, boolean includesTag,
+        boolean shouldDropBehind) throws IOException {
+    final CacheConfig writerCacheConf;
+    // Don't cache data on write on compactions.
+    writerCacheConf = new CacheConfig(store.getCacheConfig());
+    writerCacheConf.setCacheDataOnWrite(false);
+
+    InetSocketAddress[] favoredNodes = null;
+    if (store.getHRegion().getRegionServerServices() != null) {
+      favoredNodes = store.getHRegion().getRegionServerServices().getFavoredNodesForRegion(
+        store.getHRegion().getRegionInfo().getEncodedName());
```

Comment:
Why not use this directly from the StoreContext: https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreContext.java#L95

Reply:
I already need an instance of HStore, as you noted on your comment below. This HStore instance need is not exclusive from this Compactor extension, it is already defined on the parent classes single declared constructors, so it's used for other stuff in the hierarchy, not only the filereader creation. I don't think there's much value in create an extra StoreContext here when we have everything we need in HStore instance itself.

Reply:
The StoreContext is initialized with the store so there is no need to create a context. I'm specifically suggesting to use the store.getStoreContext().getFavoredNodes() instead of having to expose the RSServices and have all this extra code here.

Reply:
Wasn't finding `store.getStoreContext()` before, since it was package private. Made it public on HStore and am calling it there now.